mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,44 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
-        if (agg == null) {
-          agg = Array.fill(vec.size)(
-            new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))
-        }
-        require(vec.size == agg.length,
-          s"Number of dimensions must be ${agg.length} but got ${vec.size}")
-        var i = 0
-        while (i < vec.size) {
-          agg(i) = agg(i).insert(vec(i))
-          i += 1
-        }
-      }
+    val vectors = dataset.select($(inputCol)).rdd.map { case Row(vec: Vector) => vec }
+    val numFeatures = MetadataUtils.getNumFeatures(dataset.schema($(inputCol)))
+      .getOrElse(vectors.first().size)
 
-      if (agg == null) {
-        Iterator.empty
-      } else {
-        Iterator.single(agg.map(_.compress))
-      }
-    }.treeReduce { (agg1, agg2) =>
-      require(agg1.length == agg2.length)
-      var i = 0
-      while (i < agg1.length) {
-        agg1(i) = agg1(i).merge(agg2(i))
-        i += 1
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    // compute scale by the logic in treeAggregate with depth=2
+    // TODO: design a common treeAggregateByKey in PairRDDFunctions?
+    val scale = math.max(math.ceil(math.sqrt(vectors.getNumPartitions)).toInt, 2)
```

Comment:
Thinking this through more: say there are M partitions of data, and N dimensions / features. This produces N*sqrt(M) new partitions. 

N*sqrt(M) is better than than N before, for the case I mentioned, where N is small and M is large. The number of partitions doesn't drop so much that you might lose too much parallelism for large data. You're still losing some parallelism compared to processing with M partitions, in the case where sqrt(M) > N.

N*sqrt(M) is on the other hand a lot of partitions, if N is large and M is small, which is your case. What if N = 10000? You might go from ten partitions to a thousand. Or a million if M is large as well as N.

I'm trying to figure out if we can mostly get rid of the factor of N.

What if you keyed by (i, partition / N)? You end up with about N * M/N = M partitions in the aggregation, which is nice. Well, that's roughly true if N < M. Of course, if N > M then this still gives you N partitions out, but that's better than N*sqrt(M) for your case I believe. WDYT?

I think this still bears a quick benchmark on the 'common case' of few features and a nontrivial amount of data / partitions. I am slightly concerned this might be much slower for that case.

Reply:
> say there are M partitions of data, and N dimensions / features. This produces N*sqrt(M) new partitions.

Sorry I am try to figure out what  _partition_ here means. Does it means a intermediate aggregator `QuantileSummaries`?
Or if you mean a _partition_ in internal rdds, I think it is still M partitions, since all the ops (`mapPartitionsWithIndex`->`aggregateByKey`->`map`->`reduceByKey`) before `collect` do not change numPartitions.

The base logic is quite similar to `treeAggregate`:
Say a key (i=1), the orginal rdd has 16 partitions. There will be 16 aggregator `QuantileSummaries` for i=1 at first, one per partition. 
Then after `aggregateByKey` there will partitally merged and there are 4 `QuantileSummaries`.
Finally `reduceByKey`  merge them to generate only one `QuantileSummaries` on some executor.

Reply:
Hm yeah nevermind the comment about partitions, that doesn't change. It's just the number of aggregation keys. I don't know whether increasing the number of aggregation keys is something we should avoid or not, but that's not the reason, OK.

Anyway, I'd still be interested how it performs before and after, any rough benchmark, for the small N case. I'm just sort of concerned it could be slower for that common case because now the whole data set is chopped up finely and shuffled in all cases. Doesn't mean we can't do it, just wondering if there is a big difference to know about.

Reply:
ok, I will do some performance tests